### PR TITLE
Bandersnatch tweaks after backend update

### PR DESCRIPTION
**Pull request by davxy**
As per title

### CODE REVIEW

File: substrate/primitives/core/src/bandersnatch.rs [hunk 1]
```diff
@@ -774,51 +762,45 @@ pub mod ring_vrf {
...
 			else {
 				return false
 			};
 
-			let Ok(ring_proof) = RingProof::deserialize_compressed(self.ring_proof.as_slice())
-			else {
-				return false
-			};
+			let preouts: [bandersnatch_vrfs::VrfPreOut; N] =
+				core::array::from_fn(|i| self.outputs[i].0.clone());
```
Comment (koute) [inline substrate/primitives/core/src/bandersnatch.rs:795]: How big is the thing being cloned here? I tried to follow the types and manually see how big it is, but there are so many layers of abstraction between what's used here and the types which actually store the raw bytes (and some broken `docs.rs` builds in between :P) that I just gave up. (As far as I can see this has some inline big ints inside, so I'd like to know how big exactly.) Can you `core::mem::size_of` in a throwaway `#[test]` and paste me the result?

**Review by koute — APPROVED** [review]

File: substrate/primitives/core/src/bandersnatch.rs (hunk 1 above)
Comment (davxy) [reply substrate/primitives/core/src/bandersnatch.rs:795]: Each element is a Bandersnatch curve group element (aka curve point).
```
struct Affine {
    x: U256,            // 32 bytes
    y: U256,            // 32 bytes
    infinity: bool,
}
```

The printed size is 72 bytes (8 bytes are given to the `infinity` bool for alignment).

---

The constructed array size is bounded by the max number we support at the moment (`MAX_VRF_IOS = 3`).

Comment (davxy) [reply substrate/primitives/core/src/bandersnatch.rs:795]: @burdges probably we can implement `Copy` for `VrfInput` and `VrfPreOut`   (as the inner types are Copy)

Comment (burdges) [reply substrate/primitives/core/src/bandersnatch.rs:795]: Alright sure.  Arkworks & dalek both have copy types of this size.

In memory, yeah affine curve points are two base field elements plus a flag.  This one has a 32 byte base field, and serializes in 33 bytes.

Yes, arkworks has layers itself.  And I made it worse for other reasons.

File: substrate/primitives/core/src/bandersnatch.rs [hunk 2]
```diff
@@ -539,24 +538,26 @@ pub mod vrf {
...
+			let outputs: Vec<_> = thin_signature.preouts.into_iter().map(VrfOutput).collect();
+			let outputs = VrfIosVec::truncate_from(outputs);
+
+			let mut signature =
+				VrfSignature { signature: Signature([0; SIGNATURE_SERIALIZED_LEN]), outputs };
+
+			thin_signature
+				.proof
+				.serialize_compressed(signature.signature.0.as_mut_slice())
 				.expect("serialization can't fail");
```
[the remark this answers is not in the document]
Comment (burdges) [reply substrate/primitives/core/src/bandersnatch.rs:558]: ArkScale requires you "audit" arkworks types for this, but doing so is simply a grep command: 
https://github.com/w3f/ark-scale/blob/master/src/lib.rs#L126
https://github.com/w3f/ark-scale/blob/master/README.md

Also VrfSignature became a scale type:
https://github.com/w3f/ring-vrf/blob/master/dleq_vrf/src/traits.rs#L97

File: substrate/primitives/core/src/bandersnatch.rs [hunk 3]
```diff
@@ -741,31 +733,27 @@ pub mod ring_vrf {
...
+				RingVrfSignature { outputs, signature: [0; RING_SIGNATURE_SERIALIZED_LEN] };
 
-			let mut ring_proof = [0; RING_PROOF_SERIALIZED_LEN];
 			ring_signature
-				.ring_proof
-				.serialize_compressed(ring_proof.as_mut_slice())
-				.expect("ring-proof serialization can't fail");
+				.proof
+				.serialize_compressed(signature.signature.as_mut_slice())
+				.expect("ring-signature serialization can't fail");
```
[the remark this answers is not in the document]
Comment (davxy) [reply substrate/primitives/core/src/bandersnatch.rs]: https://github.com/paritytech/polkadot-sdk/pull/1482/commits/bccbf5a9a2583e481c5e65fbbbd1187a90547e88

I added some more assertions in the "backend sanity checks" test.
This test basically checks our assumptions about the serialized lengths of some objects.

These serialized lengths are constant and any variation of the sizes is early catched by the test

File: substrate/primitives/core/src/bandersnatch.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (burdges) [reply substrate/primitives/core/src/bandersnatch.rs]: Oh he you mean the length, yeah a test for this makes sense.

**Review by bkchr — APPROVED** [review]

**Pull request merged**